code4me-server/src/api.py

```
@@ -1,22 +1,132 @@
+
+    predictions = {model.name: prediction for model, prediction in zip(Model, predictions)}
+    return time, predictions
+
+@v2.route("/prediction/autocomplete", methods=["POST"])
+@limiter.limit("4000/hour")
+def autocomplete_v2():
+
+    try:
+        # TODO: As we want every request to be authorised, this can be extracted into a decorator
+        user_uuid = authorise(request)
+        request_json = request.json
+        
+        # current_app.logger.warning(request_json)
+        filter_time, filter_type, should_filter = filter_request(user_uuid, request_json)
+
+        predict_time, predictions = get_predictions(request_json) \
+            if (not should_filter) or (request_json['trigger'] == 'manual') \
+            else (None, {}) 
+
+        log_filter = f'\033[1m{"filter" if should_filter else "predict"}\033[0m'
+        log_context = f'{request_json["prefix"][-10:]}•{request_json["suffix"][:5]}'
+        current_app.logger.warning(f'{log_filter} {log_context} \t{filter_type} {[v[:10] for v in predictions.values()]}')
+
+        verify_token = uuid.uuid4().hex if not should_filter else ''
+        prompt_survey = should_prompt_survey(user_uuid) if not should_filter else False
+
+        store_completion_request(user_uuid, verify_token, {
+            **request_json,
+            'timestamp': datetime.now().isoformat(),
+            'filter_type': filter_type,  
+            'filter_time': filter_time,
+            'should_filter': should_filter,
+            'predict_time': predict_time,
+            'predictions': predictions,
+            'survey': prompt_survey,
+            'study_version': '0.0.1'
+        })
+
+        return {
+            'predictions': predictions,
+            'verifyToken': verify_token,
+            'survey': prompt_survey
+        }
+
+    except Exception as e:
+        # logging may be a good idea for debugging
+        traceback.print_exc()
+        return response({
+            "error": str(e)
```

Returning an error string might expose secrets; maybe return an error uuid here and log the error id along with the error in the console so it can be further inspected by staff if reported.

---

I now generate an `error_uuid` and return that to the user; while logging traceback and other info related to the `error_uuid` on the server-side.